[aotinductor] support _scaled_dot_product_flash_attention fallback

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110003

This PR supports _scaled_dot_product_flash_attention fallback kernel.
Note that in the abi_compatible mode, we retrieve outputs by passing
output argument pointers rather than relying on std::get.

It also fixes an issue related to dynamic shapes, where we wrongfully
query undefined dynamic symbols.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov

Commit:
[inductor] support _scaled_dot_product_flash_attention fallback

This PR supports _scaled_dot_product_flash_attention fallback kernel.
Note that in the abi_compatible mode, we retrieve outputs by passing
output argument pointers rather than relying on std::get.

It also fixes an issue related to dynamic shapes, where we wrongfully
query undefined dynamic symbols.

[ghstack-poisoned]

ezyang commented:
I think I will mostly defer to @desertfire here. Is there something specifically you want me to review?

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -152,6 +152,22 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
+    AtenTensorHandle query,
+    AtenTensorHandle key,
+    AtenTensorHandle value,
+    ...);
```
Variadic function in the ABI?! Seems suspicious. Tell me more?

khabinov replied:
Agree, looks kinda suspicious. Maybe we need to define something like:

```
aoti_torch__scaled_dot_product_flash_attention_6()
aoti_torch__scaled_dot_product_flash_attention_7()
aoti_torch__scaled_dot_product_flash_attention_8()
```

ezyang replied:
Naively, I would have expected to take all optional arguments and force you to fill in the defaults.

chenyang78 replied:
```torch.ops.aten._scaled_dot_product_flash_attention``` takes a couple of default arguments (dropout_p, is_causal, return_debug_mask and scale) that may not be presented in the IR. For example, given ``` torch.ops.aten._scaled_dot_product_flash_attention(q, k, v)```, at https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L3853, we only see three args for ```q```, ```k```, ```v```. We are missing all the default arguments. Consequently, at the codegen time, we don't know the default values for generating ```_scaled_doc_product_flash_attention``` fallback. (Question - is there any way that we could get the default argument values when we create IR for this FallbackKernel? )

Note that we could make default values for ```_scaled_doc_product_flash_attention``` by explicitly detecting the kernel name in the wrapper code, but I think it may be not general enough. For example, we might end up with a number of special rules for various fallback kernels in the wrapper. 

Since C doesn't support default arguments, I went with the variadic-function solution.

desertfire replied:
`ordered_kwargs_for_cpp_kernel` should give you the default values for which is read from schema.

chenyang78 replied:
> `ordered_kwargs_for_cpp_kernel` should give you the default values for which is read from schema.

Hmm, we don't use schema for ```_scaled_dot_product_flash_attention```. Do we want to use schema for it in a way similar to other kernels like ```ConvolutionUnary```?

chenyang78 replied:
> Agree, looks kinda suspicious. Maybe we need to define something like:
> 
> ```
> aoti_torch__scaled_dot_product_flash_attention_6()
> aoti_torch__scaled_dot_product_flash_attention_7()
> aoti_torch__scaled_dot_product_flash_attention_8()
> ```

Hmm, it's probably less ideal, because we would have a special rule for the ```scaled_dot_product_flash_attention``` fallback in the wrapper code to generate these.

khabinov replied:
> Naively, I would have expected to take all optional arguments and force you to fill in the defaults.

Yeah, makes sense.

chenyang78 replied:
Per discussion, replaced the variadic function with the one with default arguments being filled up. This change requires us to replace ```torch.ops.aten._scaled_dot_product_flash_attention``` with ```torch.nn.functional.scaled_dot_product_attention```. Only the latter provides a schema that can be used to retrieve the default values for the optional arguments.

ezyang reviewed (changes requested):
no variadic functions allowed unless you convince me it's a good idea

khabinov commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,186 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
+    at::Tensor* query_tensor = tensor_handle_to_tensor_pointer(query);
+    at::Tensor* key_tensor = tensor_handle_to_tensor_pointer(key);
+    at::Tensor* value_tensor = tensor_handle_to_tensor_pointer(value);
+    auto ret = at::_scaled_dot_product_flash_attention(
+        *query_tensor,
+        *key_tensor,
+        *value_tensor,
+        dropout_p,
+        is_causal,
+        return_debug_mask,
+        scale);
+
+    at::Tensor* ret0_tensor = new at::Tensor(std::move(std::get<0>(ret)));
+    *ret0 = tensor_pointer_to_tensor_handle(ret0_tensor);
+    at::Tensor* ret1_tensor = new at::Tensor(std::move(std::get<1>(ret)));
+    *ret1 = tensor_pointer_to_tensor_handle(ret1_tensor);
+    // ret2 and ret3 may be null
+    if (ret2) {
+      at::Tensor* ret2_tensor = new at::Tensor(std::move(std::get<2>(ret)));
+      *ret2 = tensor_pointer_to_tensor_handle(ret2_tensor);
+    }
+    if (ret3) {
+      at::Tensor* ret3_tensor = new at::Tensor(std::move(std::get<3>(ret)));
+      *ret3 = tensor_pointer_to_tensor_handle(ret3_tensor);
+    }
+    *ret4 = std::get<4>(ret);
+    *ret5 = std::get<5>(ret);
+    at::Tensor* ret6_tensor = new at::Tensor(std::move(std::get<6>(ret)));
+    *ret6 = tensor_pointer_to_tensor_handle(ret6_tensor);
+    at::Tensor* ret7_tensor = new at::Tensor(std::move(std::get<7>(ret)));
+    *ret7 = tensor_pointer_to_tensor_handle(ret7_tensor);
+    at::Tensor* ret8_tensor = new at::Tensor(std::move(std::get<8>(ret)));
+    *ret8 = tensor_pointer_to_tensor_handle(ret8_tensor);
```
For brevity we can use structured binding, e.g.:
```
auto [ret0, ret1, ret2, ret3, ret4, ret5, ret6, ret7, ret8] = at::_scaled_dot_product_flash_attention(
        *query_tensor,
        *key_tensor,
        *value_tensor,
        dropout_p,
        is_causal,
        return_debug_mask,
        scale);
```

khabinov commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,186 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
+    double dropout_p = 0.0,
+    bool is_causal = false,
+    bool return_debug_mask = false,
+    c10::optional<double> scale = c10::nullopt) {
```
Is it supposed to be C friendly API? Why `c10::optional` then?

chenyang78 replied:
> Is it supposed to be C friendly API? Why `c10::optional` then?

We provide C interface. The implementation can be in C++.

khabinov commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -152,6 +152,22 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
     AtenTensorHandle* ret // returns new reference
 );
 
+AOTI_TORCH_EXPORT AOTITorchError aoti_torch__scaled_dot_product_flash_attention(
```
Do we use it in this diff?

chenyang78 replied:
> Do we use it in this diff?

Yes, we do.

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,186 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
   });
 }
 
+static AOTITorchError aoti_torch__scaled_dot_product_flash_attention_internal(
```
Nit: move this to the anonymous namespace.

chenyang78 replied:
I think ```static``` makes internal linkage. But Yes, I will move it to the anon namespace for consistency.

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1402,21 +1413,53 @@ def generate_c_shim_extern_kernel_call(self, kernel, args):
+    def generate_c_shim_fallback_kernel_call(self, fallback_kernel, args):
+        output_args = []
+        output_raii_handles = []
+        output_name_base = fallback_kernel.get_name()
```
> We probably need a counter to distinguish different fallback_kernel_calls?

I think different fallback_kernel_calls have a unique buffer name, buf0, buf1, etc?

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1584,6 +1631,11 @@ def codegen_reinterpret_view(self, name, size, stride, offset, writer) -> str:
             return f"reinterpret_tensor({', '.join(args)})"
 
+    def codegen_multi_output(self, name, value):
+        # if V.graph.aot_mode and name in set(V.graph.get_output_names()):
```
It's accidentally left here. Will remove it.

chenyang78 commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -152,6 +152,22 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
+    int64_t* ret5,
+    AtenTensorHandle* ret6, // returns new reference
+    AtenTensorHandle* ret7, // returns new reference
+    AtenTensorHandle* ret8, // returns new reference
```
> Looks like we have to put return values at the beginning of parameters because of varargs, so we should just revert #109834 for consistency. We can still change at this point.

I will revert #109834 once the variadic solution is accepted.

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,186 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
+  va_start(args, value);
+
+  double dropout_p = 0.0;
+  if (num_inputs >= 4) {
```
Ah, good catch! Will fix it.

Commit:
Update on "[inductor] support _scaled_dot_product_flash_attention fallback"


This PR supports _scaled_dot_product_flash_attention fallback kernel.
Note that in the abi_compatible mode, we retrieve outputs by passing
output argument pointers rather than relying on std::get.

It also fixes an issue related to dynamic shapes, where we wrongfully
query undefined dynamic symbols.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,62 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
+            return_debug_mask,
+            scale);
+
+    at::Tensor* ret0_tensor = new at::Tensor(std::move(r0));
```
I know I have been using this in other places, but now seems a good time to write a utility function for this `new` + `tensor_pointer_to_tensor_handle` pattern.

chenyang78 replied:
Yes, will do it in a follow-up PR. Thanks.

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -599,6 +606,7 @@ def is_expr(x):
             for dim, shape in enumerate(shapes):
                 shape = V.graph.sizevars.simplify(shape)
                 if shape in needed:
+                    self.defined_symbols.add(shape)
```
This seems wrong.  `shape` might be `s0+s1+1`.

Perhaps you want `.free_symbols`?

chenyang78 replied:
```shape``` is guarded by ```needed```, which is ```needed = V.graph.sizevars.free_symbols()``` at line 576 above, so we won't generate bad code for non-symbol expressions such as ```s0 + s1```.

jansel replied:
I am worried we will miss the symbols though.

`s0+s1 in {s0}` will return false and defined symbols will be wrong.

chenyang78 replied:
@jansel The issue is addressed in https://github.com/pytorch/pytorch/pull/110411, where I removed the source of generating such bad code. There was some historical reason why we chose the previous approach. I think the changes in #110411 make the implementation much simpler and less error-prone.

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -152,6 +152,25 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
+    AtenTensorHandle* ret2, // returns new reference
+    AtenTensorHandle* ret3, // returns new reference
+    int64_t* ret4,
+    int64_t* ret5,
```
dynamic shapes?!

chenyang78 replied:
Those two fields stand for max_seqlen_batch_q and max_seqlen_batch_kv, which are returned from the sdpa kernel:

https://github.com/pytorch/pytorch/blob/36eb1bb5486e6899a806cbe82c42c1021d456e3e/aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp#L245-L246

I think they can be dynamic given they are related to sequence length and batch?

ezyang commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -152,6 +152,25 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
+    int64_t* ret5,
+    AtenTensorHandle* ret6, // returns new reference
+    AtenTensorHandle* ret7, // returns new reference
+    AtenTensorHandle* ret8 // returns new reference
```
If this is perf sensitive, you're going to want an out variant of this kernel at some point

chenyang78 replied:
Yes, good point. Thanks.

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -3627,6 +3627,10 @@ def __init__(
+        # We need output buffers for generating kernel arguments in the
+        # abi-compatible mode, where we retrieve outputs by pass each individual
+        # output through the abi-compatible interface.
+        self.outputs = []
```
Is there a reason this can't be populated at construction time?

chenyang78 replied:
Hmm, it is populated at construction time (in the classmethod ```create```):

https://github.com/pytorch/pytorch/pull/110003/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acR3886

Or, you were referring to another point?

ezyang replied:
no, I understand what this is doing now

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -290,6 +290,7 @@ def __init__(self):
         self.first_device_guard = True
         self.supports_intermediate_hooks = True
         self.expr_printer = pexpr
+        self.defined_symbols = set()
```
I still don't really understand the fix here, and I would like to understand it. Could you describe what goes wrong and how this fixes it?

chenyang78 replied:
The original issue showed up with a large model. Let me come up with a simple test to demonstrate it.

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -182,6 +185,62 @@ AOTITorchError aoti_torch_create_tensor_from_blob(
+    *ret0 = tensor_pointer_to_tensor_handle(ret0_tensor);
+    at::Tensor* ret1_tensor = new at::Tensor(std::move(r1));
+    *ret1 = tensor_pointer_to_tensor_handle(ret1_tensor);
+    // ret2 and ret3 may be null
```
These are for cases where the corresponding outputs are ```None``` (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L3883). In such cases, I chose to pass ```nullptr```, because I think it seems to match the python IR more closely? We could pass in a dummy valid pointer though. I am fine with either way. What's your preference? Thanks.

ezyang replied:
So are you saying something like, because we specialize on things being None or not, at compile time we KNOW whether or not ret2/ret3 return None, and so if they are returning None we don't have to pass in a pointer because we know it's always going to be none? I guess this is the variadic thing all over again; technically, we could have just had separate overloads for if ret2/ret3 return None or not.

I guess this as is is fine, but I'd like it if we asserted that ret2_tensor was in fact undefined if ret2 was null.

chenyang78 replied:
> So are you saying something like, because we specialize on things being None or not, at compile time we KNOW whether or not ret2/ret3 return None, and so if they are returning None we don't have to pass in a pointer because we know it's always going to be none? 

Yes, this is exactly the case.

ezyang commented:
I may moot the flash attention int returns, it looks to me that they are not actually dynamic. Confirming with @drisspg . See also https://github.com/pytorch/pytorch/issues/110322

ezyang commented:
see https://github.com/pytorch/pytorch/pull/110527

chenyang78 commented:
> see #110527

Got it. Thanks for addressing this.

ezyang commented:
We are going to leave this in but our expectation is for dense tensor the int returns will never actually get used

drisspg commented:
> We are going to leave this in but our expectation is for dense tensor the int returns will never actually get used

Question if we wanted to change this for dense tensors, and land something like this: https://github.com/pytorch/pytorch/pull/110546 (after a rebase and some error fixes). Would that be a problem, FC/BC wise?

ezyang commented:
not for aoti as this isn't deployed yet

chenyang78 commented:
Closing this in favor of https://github.com/pytorch/pytorch/pull/110085
